### ISSUE TITLE
Watchman fails with 'Operation not permitted'

### ISSUE THREAD

**Issue by orome** [facebook/watchman#1030]
Since updating yesterday using Hombrew on macOS, watchman always fails with the error:
```json
{
    "version": "2022.05.30.00",
    "error": "std::__1::system_error: open: /Path/To/Project: Operation not permitted"
}
```
Does the 2022.05.30.00 version of Watchman work? if so, why would it fail (for me at least) suddenly after being updated?

As far as I know the _only_ difference between the configurations of my machine today and yesterday (when Watchman worked fine) is that I ran `brew upgrade watchman`. I have confirmed [all the necessary file permissions are in place](https://stackoverflow.com/q/72451781/656912) and have rebooted, reinstalled, terminated, etc.

_UPDATE: [A solution](https://github.com/facebook/watchman/issues/1030#issuecomment-1145766948) to get things working while we await a fix to Watchman is provided below:_

```text
watchman shutdown-server       # To ensure an old version isn't still running
brew uninstall watchman        # Get rid of the buggy (e.g., 2022.05.30.00) version
curl https://raw.githubusercontent.com/Homebrew/homebrew-core/8651d8e23d308e564414188509f864e40548f514/Formula/watchman.rb > "$(brew --repository)/Library/Taps/homebrew/homebrew-core/Formula/watchman.rb"
brew install watchman          # This installs the functioning 2022.05.16.00 version dowloaded above
brew pin watchman              # Tell brew to leave this older version alone (don't forget to unpin once this problem is solved)
cd "$(brew --repository)/Library/Taps/homebrew/homebrew-core/Formula/"
git checkout --  watchman.rb   # To avoid brew showing warnings about having uncommitted git changes
```

Note that for those who have found other solutions, you are not experiencing the same issue, but of you are doing things like granting watchman full disk access, you may be reducing your security unnecesarily.

Also, if you are on macOS, and the above doesn't work on its own, it may pay off to do the following and try the above again:
```text
sudo tccutil reset SystemPolicyDocumentsFolder
sudo tccutil reset SystemPolicyAllFiles
```
If you do this, you will need to again grant access to those apps that need it (they will ask).

**_UPDATE:As of 2022.06.27.00 Watchman appears to work again._**

If you followed the instructions above, just 
```text
brew unpin watchman
```
and update it as usual. Then (and from then on), to be sure things work as expected, after every update of Watchman:
```text
watchman watch-del-all
watchman shutdown-server 
```

**Comment by orome** [comment]
In an attempt to fix this I've run
```bash
brew update
brew uninstall watchman
sudo tccutil reset SystemPolicyDesktopFolder
sudo tccutil reset SystemPolicyDeveloperFiles
sudo tccutil reset SystemPolicyDocumentsFolder
sudo tccutil reset SystemPolicyDownloadsFolder
sudo tccutil reset SystemPolicyNetworkVolumes
sudo tccutil reset SystemPolicyRemovableVolumes
sudo tccutil reset SystemPolicySysAdminFiles
```
then rebooted and run
```bash
brew install watchman
watchman
```
That results in a series of library errors that are cleared by reinstalling several of watchman's dependencies. Once that's done I did
```bash
brew update
brew uninstall watchman
```
then rebooted again and did
```bash
brew install watchman
watchman
```
and still get the same error:
```bash
{
    "version": "2022.05.30.00",
    "error": "std::__1::system_error: open: /Path/To/Project: Operation not permitted"
}
```

**Comment by beamcode** [comment]
same issue wtf

**Comment by orome** [comment]
@beamcode Really vexing bug. I basically redid all my permissions to get rid of it. It seems to be part of a longer running problem with watchman: I had multiple (about 8) entries for watchman in my Files & Folders settings, all for executable versions that were no longer present (as indicated by the lack of an icon) and that were causing issues with getting settings to stick.

So there's some kind of bad behavior by watchman relating to File & Folders permissions — and possibly how it identifies itself to macOS — that has been going on for a while, it appears.

**Comment by beamcode** [comment]
https://i.ibb.co/8X5N1tL/Screenshot-2022-06-01-at-23-05-04.png
i manually added watchman to the full disk access what else can i do i reinstall like 5 times
Maybe its something to do with a recent macos version because i just got my new m1 mac like 3 days ago and updated everything to latest updates idk

**Comment by calluthom** [comment]
The only thing that seems to work is moving your entire project to your Home folder. Nothing else seems to work. It seems to be affecting a lot of other users right now, likely a bug with some new macos security update with file access permissions.

**Comment by orome** [comment]
> I have the same error after following the React Native getting started guide, on both xcode and expo cli.

Yes, this bug kills `expo start`: no development is happening here.

**Comment by woyounai** [comment]
guys,you need reinstall "version": "2022.05.16.00".

**Comment by orome** [comment]
> guys,you need reinstall "version": "2022.05.16.00".

I tried that both using the ZIP (which doesn't work at all #1031) and using previous Hombrew formulas, none of which produced a functioning installation. How did you manage to install a previous version?

**Comment by beamcode** [comment]
> The only thing that seems to work is moving your entire project to your Home folder. Nothing else seems to work. It seems to be affecting a lot of other users right now, likely a bug with some new macos security update with file access permissions.
> 
> I have the same error after following the React Native getting started guide, on both xcode and expo cli.
> 
> I hope the watchman team looks into this!

i think he's right its probably a security update thing

**Comment by orome** [comment]
> The only thing that seems to work is moving your entire project to your Home folder.

Confirming that this is also the only thing that works for me. Not practical solution, but it will allow some headway to be made.

**Comment by gorhom** [comment]
same issue here, it only worked when moving the project to the home folder 😕

MacOS 12.4 ( M1 )

**Comment by oetjen** [comment]
This is definetly a problem with version 2022.05.30.00.
If you rollback to 2022.05.16.00, it works again.
Howto: See https://remarkablemark.org/blog/2017/02/03/install-brew-package-version/

TLDR: 
```
brew uninstall watchman
curl https://raw.githubusercontent.com/Homebrew/homebrew-core/8651d8e23d308e564414188509f864e40548f514/Formula/watchman.rb > /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/watchman.rb
brew install watchman
brew pin watchman # Don't forget to unpin once this problem is solved...
git checkout --  /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/watchman.rb
```

**Comment by gorhom** [comment]
> This is definetly a problem with version 2022.05.30.00. If you rollback to 2022.05.16.00, it works again. Howto: See https://remarkablemark.org/blog/2017/02/03/install-brew-package-version/
> 
> TLDR:
> 
> ```
> brew uninstall watchman
> curl https://raw.githubusercontent.com/Homebrew/homebrew-core/8651d8e23d308e564414188509f864e40548f514/Formula/watchman.rb > /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/watchman.rb
> brew install watchman
> brew pin watchman # Don't forget to unpin once this problem is solved...
> git checkout --  /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/watchman.rb
> ```

thanks @oetjen! your solution did fix the issue 🙌

**Comment by orome** [comment]
@oetjen I get:
```text
/opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/watchman.rb: No such file or directory
```
I don't even have an `/opt/homebrew/`, just an `/opt/homebrew-cask/`, which is empty.

**Comment by shoopi12** [comment]
2022.05.30.00 appears to be broken.

Downgrading to a previous version is surprisingly not easy.

Decided to save some time and move the project from Documents folder to Documents/Projects and it worked.
Will bring it back once the bug is fixed I suppose.

**Comment by orome** [comment]
> Downgrading to a previous version is surprisingly not easy.

@shoopi12 I recommend giving it a shot. The instructions above work. The only things to be careful of are:

1. If you are on an Intel Mac use
    ```text
    /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/
    ```
   for the Formula path.
2. Be sure to `cd` to that directory before the `git checkout` step.

**Comment by shoopi12** [comment]
@orome tried that, I'm getting the same error with v2022.05.16.00

**Comment by shoopi12** [comment]
@orome tried that, but for some reason I still get the error with the higher version:
![Screen Shot 2022-06-03 at 22 28 15](https://user-images.githubusercontent.com/25888162/171936260-16719164-b70b-45af-9bd4-fc41bb54d33c.png)

while watchman -v shows the lower version:
![Screen Shot 2022-06-03 at 22 28 26](https://user-images.githubusercontent.com/25888162/171936337-86b6fabd-f88b-4d9c-b72f-5d1432b604d4.png)

What I did:
```
brew uninstall watchman

curl https://raw.githubusercontent.com/Homebrew/homebrew-core/8651d8e23d308e564414188509f864e40548f514/Formula/watchman.rb > /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/watchman.rb

brew pin watchman

cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/watchman.rb

git checkout
```

**Comment by orome** [comment]
> @orome tried that, but for some reason I still get the error with the higher version...`

You omitted `brew install watchman`!

**Comment by shoopi12** [comment]
@orome sorry I omitted it from my Comment, I did the step of course

**Comment by liquidvisual** [comment]
None of this worked for me on an Intel iMac running Monterey. Restarted so many times.

How do you give watchman Full Disk Access, what are the steps?

**Comment by StorkZ** [comment]
I have just installed 2022.06.06.00 and the same problem still persists. 
Currently I am trying to get it to work using one of the solutions above.

**Comment by neoswallow** [comment]
> Can someone explain why the last step git checkout -- watchman.rb is required?

It is to reset the change you just made by replacing watchman.rb with a previous version.
Brew update will break if you have un-committed changes.

**Comment by zhangheng0317** [comment]
I use 2022.06.06.00, its OK

**Comment by l0gicgate** [comment]
@zhangheng0317  `2022.06.06.00` doesn't work for me. Same bug still.

**Comment by orome** [comment]
> I use 2022.06.06.00, its OK

@zhangheng0317 Were you having the issue described in the OP (with `2022.05.30`)?

**Comment by TwanLuttik** [comment]
I can also confirm this is an issue after a clean factory reset on my macbook m1 max with watchman version `2022.06.06.00`.

**Comment by DaniloPootaren** [comment]
I can confirm also that issue still persists with M1 chips. I've tried all the solutions given above and I could not resolve it.

**Comment by neoswallow** [comment]
The workaround worked around initially but it's broken again and now I'm getting different error.
```
app/node_modules/metro-hermes-compiler/src/emhermesc.js:77
          throw ex;
          ^

Error: EMFILE: too many open files, watch
    at FSEvent.FSWatcher._handle.onchange (node:internal/fs/watchers:204:21)
Emitted 'error' event on NodeWatcher instance at:
    at NodeWatcher.checkedEmitError (app/node_modules/sane/src/node_watcher.js:143:12)
    at FSWatcher.emit (node:events:527:28)
    at FSEvent.FSWatcher._handle.onchange (node:internal/fs/watchers:210:12) {
  errno: -24,
  syscall: 'watch',
  code: 'EMFILE',
  filename: null
}
error Command failed with exit code 7.
```

And I also get this error trying to run any watchman command directly:
```
watchman shutdown-server                                                                                               
dyld[82779]: Symbol not found: __ZN5folly19shared_mutex_detail25getMaxDeferredReadersSlowERNSt3__16atomicIjEE
  Referenced from: /opt/homebrew/Cellar/watchman/2022.05.16.00/bin/watchman
  Expected in: /opt/homebrew/Cellar/folly/2022.06.06.00/lib/libfolly.0.58.0-dev.dylib
[1]    82779 abort      watchman shutdown-server
```

And I did run `brew upgrade` multiple times since the workaround steps.
Maybe we need pin more than just watchman itself.

**Comment by neoswallow** [comment]
It is possible I had folly@2022.05.30.00 when I reverted watchman to 2022.05.16.00.
I was going to revert folly to previous version but it's depend of other packages and I didn't want to break other packages to fix watchman.
Redoing workaround steps didn't help.
So it is possible this workaround steps may not help anyone else who tried after folly (or fb303 or any other watchman dependencies) got updated to a version that isn't compatible with watchman@2022.05.16.00.
I had no option but to move the repo from external drive to somewhere in home directory after which metrobundler is now working again.

**Comment by kieuquang1110tt** [comment]
MacOs Monterey 12.4 Still got this problem. Watchman 2022.06.06.00. Pls help.

**Comment by Mixpeal** [comment]
> MacOs Monterey 12.4 Still got this problem. Watchman 2022.06.06.00. Pls help.

It's weird, but just

`brew uninstall watchman`

worked for me.

**Comment by orome** [comment]
FWIW, things have been working consistently for me with pinned watchman 2022.05.16.00 (only) since following the [steps above](https://github.com/facebook/watchman/issues/1030#issuecomment-1145766948), with no other packages or dependencies pinned.

**Comment by kbeliauski** [comment]
does anyone know what is the problem? I keep getting this:

Error: EMFILE: too many open files, watch
    at FSEvent.FSWatcher._handle.onchange (internal/fs/watchers.js:178:28)

Thank you!

**Comment by robert-go** [comment]
> This is definetly a problem with version 2022.05.30.00. If you rollback to 2022.05.16.00, it works again. Howto: See https://remarkablemark.org/blog/2017/02/03/install-brew-package-version/
> 
> TLDR:
> 
> ```
> brew uninstall watchman
> curl https://raw.githubusercontent.com/Homebrew/homebrew-core/8651d8e23d308e564414188509f864e40548f514/Formula/watchman.rb > /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/watchman.rb
> brew install watchman
> brew pin watchman # Don't forget to unpin once this problem is solved...
> cd /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/
> git checkout --  watchman.rb
> ```

`➜  ~ watchman -v
dyld[3623]: Symbol not found: (__ZN5folly19shared_mutex_detail25getMaxDeferredReadersSlowERNSt3__16atomicIjEE)
  Referenced from: '/opt/homebrew/Cellar/watchman/2022.05.16.00/bin/watchman'
  Expected in: '/opt/homebrew/Cellar/folly/2022.06.06.00/lib/libfolly.0.58.0-dev.dylib'
[1]    3623 abort      watchman -v`

**Comment by robert-go** [comment]
it's not working for me.

**Comment by legion-zver** [comment]
@robert-go try adding "sh" permissions to the disk

**Comment by TwanLuttik** [comment]
> @robert-go try adding "sh" permissions to the disk

What is the file location for sh _(I'm on mac btw)_

**Comment by musasoftlabx** [comment]
This fix will work. Kindly try it

Install latest watchman version (brew install watchman)

1. Go to System preferences > Security & Privacy then scroll down
2. Select Full Disk Access
3. Unlock it and click + button to add new app/tool
4. Select Macintosh HD in the folders list. (For new users, if u cant see this folder, fo to finder preferences > sidebar and check it.)
5. Press this combination shift Command . (don't forget the dot) at the same time to see all the directories list
6. Click opt > homebrew > bin (Tested only on MacOS Monterey. If using another version, search watchman under Macintosh HD).
7. Look for watchman in the list then select it.
8. Once added, run ur program again.

Make sure ur project folder is under /Users/<username>/<project>. Don't put it under documents or any other folder there. It will refuse due to some weird restrictions.

**Comment by Seunope** [comment]
`brew uninstall watchman`

Then I ran the program again. In my case `npm run android`. This resolved the issue.

**Comment by chadaustin** [comment]
I suspect this regression was introduced by https://github.com/facebook/watchman/commit/db90c1bfc87c76d5169ed4958b778efa2e82c188

We're working on a more targeted fix to the original issue, but in the meantime hope to revert the offending change.

Until then, @musasoftlabx's fix that grants Watchman Full Disk Access is worth trying.

**Comment by legion-zver** [comment]
> > @robert-go try adding "sh" permissions to the disk
> 
> What is the file location for sh? _(I'm on mac btw)_

```bash
which sh
```
/bin/sh

**Comment by nathanaelphilip** [comment]
@chadaustin any updates?

**Comment by Ebeldev** [comment]
@musasoftlabx 

What are the weird restrictions, you are talking about?

I did follow all the steps you mentionned.

All my react-native projects are under the documents folder. Never got any problem since 5 years.

**Comment by musasoftlabx** [comment]
@Ebeldev , tested mine with M1 Pro with Monterey and got errors when I placed my project under /Documents.
I think previous versions have no issue. Am guessing the OS restricts those folders unless you chmod to change permissions

**Comment by orome** [comment]
@musasoftlabx This issue also affects Intel Macs running macOS 11.6.6.

**Comment by Ebeldev** [comment]
@musasoftlabx Thanks for the explaination. 

I'm running on Intel, with Monterrey. 

I think thoses permissions does affect my projects setup not only reactnative. I have been experiencing problems with sockets on mariadb since I upgraded macos, brew.

**Comment by chadaustin** [comment]
> @chadaustin any updates?

Backout in https://github.com/facebook/watchman/commit/cf0c3bdec32f2f457a0b06fcd1c09a62c11cf143

We don't maintain the homebrew build, but hopefully a new one comes out soon.

**Comment by TronNatthakorn** [comment]
Update 2022-06-16 This work for me. (Mac OS Monterey 12.04 - Intel Chip)

`watchman watch-del-all`
`watchman shutdown-server`
`brew unpin watchman`
`brew uninstall watchman`
`curl https://raw.githubusercontent.com/Homebrew/homebrew-core/8651d8e23d308e564414188509f864e40548f514/Formula/watchman.rb > /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/watchman.rb`
`brew install watchman`
`brew pin watchman`
`cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/`
`git checkout --  watchman.rb`

And restart your mac. 

I hope watchman team will fix this soon.

**Comment by ronakmoussavi** [comment]
Mac OS Monterey 12.4 - M1 Chip

The solutions mentioned above haven't worked for me... 

Now I'm getting this in metro:
node:events:505
      throw er; // Unhandled 'error' event
      ^

Error: EMFILE: too many open files, watch
    at FSWatcher._handle.onchange (node:internal/fs/watchers:204:21)
Emitted 'error' event on NodeWatcher instance at:
    at NodeWatcher.checkedEmitError (/path/to/project/node_modules/sane/src/node_watcher.js:143:12)
    at FSWatcher.emit (node:events:527:28)
    at FSWatcher._handle.onchange (node:internal/fs/watchers:210:12) {
  errno: -24,
  syscall: 'watch',
  code: 'EMFILE',
  filename: null
}

Any idea?

**Comment by Ebeldev** [comment]
@TronNatthakorn  This did not work for me. 

Mac Intel Monterey 12.4

Still got this error: 

jest-haste-map: Watchman crawl failed. Retrying once with node crawler.
  Usually this happens when watchman isn't running. Create an empty `.watchmanconfig` file in your project's root folder or initialize a git or hg repository in your project.
  Error: Watchman error: std::__1::system_error: open: /Users/etiennebelanger/Documents/apps/react-native-apps/rabaischocs-app: Operation not permitted. Make sure watchman is running for this project. See https://facebook.github.io/watchman/docs/troubleshooting.
/Users/etiennebelanger/Documents/apps/react-native-apps/rabaischocs-app/node_modules/metro-hermes-compiler/src/emhermesc.js:77
          throw ex;
          ^

Error: std::__1::system_error: open: /Users/etiennebelanger/Documents/apps/react-native-apps/rabaischocs-app: Operation not permitted
    at BunserBuf.<anonymous> (/Users/etiennebelanger/Documents/apps/react-native-apps/rabaischocs-app/node_modules/fb-watchman/index.js:95:23)
    at BunserBuf.emit (events.js:400:28)
    at BunserBuf.process (/Users/etiennebelanger/Documents/apps/react-native-apps/rabaischocs-app/node_modules/bser/index.js:292:10)
    at /Users/etiennebelanger/Documents/apps/react-native-apps/rabaischocs-app/node_modules/bser/index.js:247:12
    at processTicksAndRejections (internal/process/task_queues.js:77:11)
Emitted 'error' event on WatchmanWatcher instance at:
    at Client.<anonymous> (/Users/etiennebelanger/Documents/apps/react-native-apps/rabaischocs-app/node_modules/jest-haste-map/build/lib/WatchmanWatcher.js:172:10)
    at Client.emit (events.js:400:28)
    at BunserBuf.<anonymous> (/Users/etiennebelanger/Documents/apps/react-native-apps/rabaischocs-app/node_modules/fb-watchman/index.js:107:12)
    at BunserBuf.emit (events.js:400:28)
    at /Users/etiennebelanger/Documents/apps/react-native-apps/rabaischocs-app/node_modules/bser/index.js:249:12
    at processTicksAndRejections (internal/process/task_queues.js:77:11) {
  watchmanResponse: {
    error: 'std::__1::system_error: open: /Users/etiennebelanger/Documents/apps/react-native-apps/rabaischocs-app: Operation not permitted',
    version: '2022.06.13.00'
  }
}

**Comment by Ro-Mou** [comment]
> On Mac OS Monterey 12.4 - M1 Chip
> 
> The solutions mentioned above haven't worked for me...
> 
> Now I'm getting this in metro: node:events:505 throw er; // Unhandled 'error' event ^
> 
> Error: EMFILE: too many open files, watch at FSWatcher._handle.onchange (node:internal/fs/watchers:204:21) Emitted 'error' event on NodeWatcher instance at: at NodeWatcher.checkedEmitError (/path/to/project/node_modules/sane/src/node_watcher.js:143:12) at FSWatcher.emit (node:events:527:28) at FSWatcher._handle.onchange (node:internal/fs/watchers:210:12) { errno: -24, syscall: 'watch', code: 'EMFILE', filename: null }
> 
> Any idea?

UPDATE: I ended up having to move the project to root directory. Worked.

**Comment by TronNatthakorn** [comment]
> @TronNatthakorn This did not work for me.
> 
> Mac Intel Monterey 12.4
> 
> Still got this error:
> 
> Operation not permitted', version: '2022.06.13.00' } }

@Ebeldev from your error log last line showed version '2022.06.13.00' . 
Please try `watchman version`
It should be `2022.05.16.00`. If you follow my way. 

Please try start metro at least twice time. Error log should show  2022.05.16.00 instead of 2022.06.13.00
Hope this help.

**Comment by neoswallow** [comment]
For those of you who said `2022.05.16.00` doesn't work, when you install watchman again after the pulling previous rb file, do check that brew install shows `2022.05.16.00`, not latest version.
For previous attempts when the workaround worked, initial attempt installed latest rather than `2022.05.16.00`.
So I uninstall again, downloaded rb file again and then installing again installed correct version.
This is just for those who didn't check which version got installed.

Unfortunately, something broken on my local that even `2022.05.16.00` doesn't work anymore and I get some other error I posted earlier.

Fortunately, latest version does work as long as I keep my repo under home directory.

I previously had my repo on external SSD.
I'm running latest macOS on M1x mac and I do brew upgrade a few times a week which is probably why `2022.05.16.00` workaround broke on my local 🤷🏼‍♂️.

**Comment by orome** [comment]
> Fortunately, latest version does work as long as I keep my repo under home directory.

@neoswallow That has always been the case, regardless of version. The problem is specifically about projects **in** the Documents folder.

**Comment by orome** [comment]
> Brew update will break if you have un-committed changes.

@neoswallow In what way does it "break"? It simply reports uncommitted changes, no?

**Comment by neoswallow** [comment]
> > Fortunately, latest version does work as long as I keep my repo under home directory.
> 
> @neoswallow That has always been the case, regardless of version. The problem is specifically about projects **in** the Documents folder.

I don't think it's specific to Documents folder as I was having exact same problem with the projects in External SSD.
I actually don't put the project in Documents folder anymore after loosing a ton of iCloud available space possibly caused by constant delete/recreate of dependency package folders like node_modules.



> > Brew update will break if you have un-committed changes.
> 
> @neoswallow In what way does it "break"? It simply reports uncommitted changes, no?

Since the file is in a git repo and I assume a part of brew update steps is issuing `git pull` command for the update.
`git pull` simply will fail if there's uncommitted changes for the file it's pulling.
Some GUI handle it by stashing the change prior to pull and pop the stash afterward.

I guess whether it will break or not depends on how smart brew update process is.
But since the workaround is not an expected user behavior so I wouldn't be surprised if brew update doesn't handle it.
But at the same time, I don't know how they handle `pin` so it was just a guess on basic git operation, not based on actual brew update process 🤷🏼‍♂️.
I would guess someone who knows how brew update + pin works can chime in and tell us if that last step is actually needed and if so, why it's needed.

**Comment by StruanMclean** [comment]
> git checkout --  watchman.rb

Thanks so much for this

**Comment by corymharper** [comment]
I've done the above solution to revert to a working version, but it seems incompatible with another tool, folly which is at a later version. Any fix there? @robert-go seemed to be having the same issue above.

**Comment by neoswallow** [comment]
That is the problem I ran into so I was going to also downgrade folly but it's used by many other packages some of which also used by other packages, etc. So I decided not to tinker further.

**Comment by corymharper** [comment]
> That is the problem I ran into so I was going to also downgrade folly but it's used by many other packages some of which also used by other packages, etc. So I decided not to tinker further.

I was able to work around this by using `brew install --HEAD watchman`

**Comment by orome** [comment]
Does [/releases/v2022.06.20.00](https://github.com/facebook/watchman/releases/tag/v2022.06.20.00) fix this?

(This version is now [distributed by Homebrew](https://formulae.brew.sh/formula/watchman).)

**Comment by Andarius** [comment]
I'm having the same issue on an external HD, 

```bash
watchman watch /Volumes/SSD_Samsung/Projects/myapp
{
    "version": "2022.05.16.00",
    "error": "watchman::RootResolveError: RootResolveError: unable to resolve root /Volumes/SSD_Samsung/Projects/myapp: RootResolveError: realpath(/Volumes/SSD_Samsung/Projects/myapp) -> Operation not permitted"
}
```

**Comment by nathanaelphilip** [comment]
I’m still getting a `Operation not permitted` on version `2022.06.20.00`. My app is on an External HDD.

**Comment by jeremyspritelyco** [comment]
I’m still getting a `Operation not permitted` on version `2022.06.27.00`. My Mac run out of space and I can not locate React Native project inside home folder.
watchman sucks.

**Comment by cktaylor** [comment]
Version `2022.06.20.00` did not have the fix in it, but from what I can tell `2022.06.27.00` does.

I had to make sure I didn't have an older watcher still around (`watchman shutdown-server`) but after doing so, the first post-upgrade watchman I triggered asked for access to files in the Documents folder - and once allowed,  worked fine without any "Operation not permitted" errors occuring.

<img width="290" alt="Screen Shot 2022-06-28 at 11 58 21 AM" src="https://user-images.githubusercontent.com/554591/176466740-ed855bda-2dda-42ae-a70e-8944e4e8500e.png">

**Comment by orome** [comment]
> Version `2022.06.20.00` did not have the fix in it, but from what I can tell `2022.06.27.00` does.

Hope that's true! Waiting for confirmation from others. 

It would be nice if a contributor could confirm and close.

**Comment by shimo44** [comment]
After following the post by @cktaylor and getting off 6.20 my issue was resolved 😮‍💨

**Comment by orome** [comment]
> After following the post by @cktaylor and getting off 6.20 my issue was resolved 

So just to be clear, you're saying that 2022.06.27.00 is working for you (where previous versions had the error described in the OP)? What chip tech and OS?

**Comment by neoswallow** [comment]
`2022.06.27.00` indeed does fix the issue for the repo on the external drive on my system (Apple Silicon/macOS 12.4).

**Comment by shimo44** [comment]
> > After following the post by @cktaylor and getting off 6.20 my issue was resolved
> 
> So just to be clear, you're saying that 2022.06.27.00 is working for you (where previous versions had the error described in the OP)? What chip tech and OS?

Sorry! Yes - 06.27 was the version I updated to that resolved this issue for me after uninstalling watchman 06.20. I also did the suggested check and ran the shutdown server command and to my surprise even after uninstalling watchman,  06.20 was still running on my machine. My workspace is a MacBook Pro M1 running Monterey 12.0.1.

**Comment by skvost** [comment]
On mac the issue for me was that I didnt approve permissions for folder to watchman. So System preferences -> Security & Privacy -> Privacy -> Files and Folders -> Watchman -> Desktop Folder.

**Comment by KingAmo** [comment]
any news here?

**Comment by oetjen** [comment]
It looks like watchman: stable 2022.08.15.00 seems to work fine now on Mac M1. You are asked and need to grant access to the folders being synched.

**Comment by ranaalisaeed** [comment]
> Version `2022.06.20.00` did not have the fix in it, but from what I can tell `2022.06.27.00` does.
> 
> I had to make sure I didn't have an older watcher still around (`watchman shutdown-server`) but after doing so, the first post-upgrade watchman I triggered asked for access to files in the Documents folder - and once allowed, worked fine without any "Operation not permitted" errors occuring.
> 
> <img alt="Screen Shot 2022-06-28 at 11 58 21 AM" width="290" src="https://user-images.githubusercontent.com/554591/176466740-ed855bda-2dda-42ae-a70e-8944e4e8500e.png">

This works, however in my case I accidentally hit "Don't Allow" and wasn't able to trigger that prompt again. I have Mac OS Monterey.  In order to allow access, go to Preferences > Security & Privacy > Privacy > Files and Folders and add tick next to Watchman > Documents Folder

**Comment by IrynaBt** [comment]
Click on the Apple Icon and open System preferences -> Click on Security & Privacy -> File & Folders -> check Watchman

**Comment by orome** [comment]
@IrynaBt That's how it's supposed to work. This is about an older version where that was not working.

**Comment by Thetiso** [comment]
> > Version `2022.06.20.00` did not have the fix in it, but from what I can tell `2022.06.27.00` does.
> > I had to make sure I didn't have an older watcher still around (`watchman shutdown-server`) but after doing so, the first post-upgrade watchman I triggered asked for access to files in the Documents folder - and once allowed, worked fine without any "Operation not permitted" errors occuring.
> > <img alt="Screen Shot 2022-06-28 at 11 58 21 AM" width="290" src="https://user-images.githubusercontent.com/554591/176466740-ed855bda-2dda-42ae-a70e-8944e4e8500e.png">
> 
> This works, however in my case I accidentally hit "Don't Allow" and wasn't able to trigger that prompt again. I have Mac OS Monterey. In order to allow access, go to Preferences > Security & Privacy > Privacy > Files and Folders and add tick next to Watchman > Documents Folder

macos 13.1 just fxuck me!

here is my solution:
1. brew update
2. brew upgrade watchman  -> 2022.12.26
3. xcode run my rn project -> fail
4. sudo tccutil reset  XXX
5. brew uninstall watchman
6. brew install watchman
7. restart mac ->  when you start mac, watchman will ask access to file&foler, YES
8. restart rn project  -> SUCCESS

**Comment by xtealer** [comment]
Seems that the following does the trick:

1.  `watchman shutdown-server`
2. watchman will ask access to file and folder click `YES`

**Comment by orome** [comment]
@xtealer Does this relate to the original issue? That was fixed since at least 2022.06.27.00.

**Comment by xtealer** [comment]
> @xtealer Does this relate to the original issue? That was fixed since at least 2022.06.27.00.

I've had this issue today on latest version `2023.01.23.00`. I had rejected it so it is why happened in my case.

**Comment by orome** [comment]
@xtealer What you describe is just a method for making sure the currently installed version is running. 

If that is sufficient to get things working, then you don't have the original issue, which was that the (then) current version produced the error described in the OP.

**Comment by Dhruvpatel32** [comment]
https://www.rich-informatik.de/blog/how-to-fix-react-native-watchman-operation-not-permitted

it solved my problem ,maybe could help you

**Comment by oetjen** [comment]
> https://www.rich-informatik.de/blog/how-to-fix-react-native-watchman-operation-not-permitted
> 
> it solved my problem ,maybe could help you

That solution is exactly describing what has already been posted right here in this thread in June 2022, please don't repeat: https://github.com/facebook/watchman/issues/1030#issuecomment-1145766948

**Comment by imhazige** [comment]
This is worked for me:
Go to Privacy and Security > Files and Folders > Watchman
Enable the "Folders" that are present under the watchman.
